core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultTopologicalOrderGraph.java

```diff
@@ -0,0 +1,152 @@
+package ai.timefold.solver.core.impl.domain.variable.declarative;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.PrimitiveIterator;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import ai.timefold.solver.core.impl.util.CollectionUtils;
+
+public class DefaultTopologicalOrderGraph implements TopologicalOrderGraph {
+    private final int[] ord;
+    private final Map<Integer, List<Integer>> componentMap;
+    private final Set<Integer>[] forwardEdges;
+    private final Set<Integer>[] backEdges;
+
+    @SuppressWarnings({ "unchecked" })
+    public DefaultTopologicalOrderGraph(final int size) {
+        this.ord = new int[size];
+        this.componentMap = CollectionUtils.newLinkedHashMap(size);
+        this.forwardEdges = new Set[size];
+        this.backEdges = new Set[size];
+        for (int i = 0; i < size; i++) {
+            forwardEdges[i] = new HashSet<>();
+            backEdges[i] = new HashSet<>();
+            ord[i] = i;
+        }
+    }
+
+    @Override
+    public void addEdge(int from, int to) {
+        forwardEdges[from].add(to);
+        backEdges[to].add(from);
+    }
+
+    @Override
+    public void removeEdge(int from, int to) {
+        forwardEdges[from].remove(to);
+        backEdges[to].remove(from);
+    }
+
+    @Override
+    public PrimitiveIterator.OfInt nodeForwardEdges(int from) {
+        return componentMap.get(from).stream()
+                .flatMap(member -> forwardEdges[member].stream())
+                .mapToInt(Integer::intValue)
+                .distinct().iterator();
+    }
+
+    @Override
+    public boolean isLooped(LoopedTracker loopedTracker, int node) {
+        return switch (loopedTracker.status(node)) {
+            case UNKNOWN -> {
+                if (componentMap.get(node).size() > 1) {
+                    loopedTracker.mark(node, LoopedStatus.LOOPED);
+                    yield true;
+                }
+                for (var backEdge : backEdges[node]) {
+                    if (isLooped(loopedTracker, backEdge)) {
+                        loopedTracker.mark(node, LoopedStatus.LOOPED);
+                        yield true;
+                    }
+                }
+                loopedTracker.mark(node, LoopedStatus.NOT_LOOPED);
+                yield false;
+            }
+            case NOT_LOOPED -> false;
+            case LOOPED -> true;
+        };
+    }
+
+    @Override
+    public int getTopologicalOrder(int node) {
+        return ord[node];
+    }
+
+    @Override
+    public void endBatchChange() {
+        AtomicInteger index = new AtomicInteger(1);
+        AtomicInteger stackIndex = new AtomicInteger(0);
```

Comment:
I think I introduced a `MutableInteger` a while back. Since you don't need atomicity, you can use that one.